torch/testing/_internal/common_distributed.py

```diff
@@ -952,6 +943,65 @@ def perThreadSetUp(self):
     def perThreadTearDown(self):
         pass
 
+    @classmethod
+    def _run_test_with_mt_pg(cls, name, timeout, world_size, callback):
+        # Show full C++ stacktraces when a Python error originating from C++ is raised.
+        os.environ["TORCH_SHOW_CPP_STACKTRACES"] = "1"
+        failed_ranks = run_with_threaded_pg(world_size, timeout, callback)
+
+        # Print based on exceptions raised from threads
+        #   SkipTest: print info for each thread
+        #   TimeoutError: raise RuntimeError for any timed out thread
+        #   Normal Exception: print error for each thread that raises exception
+        #   and raise a RuntimeError
+        error_msg = ""
+        timeout_msg = ""
+        skip_code = -1
+        for rank, exc_info in failed_ranks:
+            exc = exc_info[1]
+            if isinstance(exc, unittest.SkipTest):
+                logger.info(
+                    f"Thread {rank} skipping test {name} for following reason: {str(exc)}"
+                )
+                if skip_code < 0:
+                    skip_code = TEST_SKIPS["generic"].exit_code
+            elif isinstance(exc, TimeoutError):
+                msg = "Thread {} terminated or timed out after {} seconds\n".format(
```

Comment:
is there going to be other time out messages? shall we just make this the default?

Reply:
This is the only place to draft a timeout message. Do you mean that we have something like `error_msg = "...".format()` at the beginning of this function?